lib/moab/storage_repository.rb

```diff
@@ -139,6 +169,57 @@ def store_new_object_version(druid, bag_pathname )
       new_version
     end
 
-  end
+    private
+
+    def check_sub_dirs(sub_dir_count, sub_dirs, results, dir=nil)
+    # assuming case statements are more expensive than if-elsif
+      if sub_dir_count.zero?
+        results << result_hash(EMPTY)
+      elsif sub_dir_count > 2
+        found_unexpected(sub_dirs, results, dir)
+      elsif sub_dir_count < 2
+        missing_data(sub_dirs, results, dir)
+      elsif sub_dir_count == 2
+        expected_dirs(sub_dirs, results, dir)
+      end
+      results.flatten
+    end
+
+    def list_sub_dirs(path)
+      Dir.entries(path).select { |entry| File.join(path, entry) unless /^\..*/ =~ entry }
+    end
 
+    def found_unexpected(array, results, dir=nil)
+      required_sub_dirs = sub_dir(dir)
+      unexpected = (array - required_sub_dirs).pop
+      results << result_hash(EXTRA_DIR_DETECTED, unexpected)
+    end
+
+    def missing_data(array, results, dir=nil)
+      required_sub_dirs = sub_dir(dir)
+      missing = (required_sub_dirs - array).pop
+      results << result_hash(MISSING_DIR, missing)
+    end
+
+    def expected_dirs(array, results, dir=nil)
+      required_sub_dirs = sub_dir(dir)
+      results << result_hash(CORRECT_DIR) if array == required_sub_dirs
+    end
+
+    def result_hash(response_code, addl=nil)
+      { response_code => result_code_msg(response_code, addl) }
+    end
+
+    def result_code_msg(response_code, addl=nil)
+      format(RESPONSE_CODE_TO_MESSAGES[response_code], addl: addl)
+    end
+
+    def sub_dir(dir=nil)
```

Comment:
I would sooner have EXPECTED_DATA_SUBDIRS = ["content", "metadata"].freeze  and EXPECTED_VERSION_SUBDIRS = ["data", "manifests"].freeze  than a method like this.  I think maybe it would be better for now to split/repeat the code currently in check_sub_dirs into  check_version_dir_content and check_data_dir_content and check_metadata_dir_content  ... this is feeling a little hard to follow / a little too clever - hard to make a change for one case without impacting the other case.

Reply:
Yeah i do like the idea of splitting it out into separate methods like: ``` check_ver_dir then check_data_dir , check_manifest_dir, check_content_dir, and check_metadata_dir ``` - i can issue a ticket for this. So for now are you saying get rid of the ``` sub_dir ``` method all together and implement this logic to either pick ``` EXPECTED_DATA_SUBDIRS = ["content", "metadata"].freeze or  EXPECTED_VERSION_SUBDIRS = ["data", "manifests"].freeze ``` within each method that was calling ``` sub_dir ``` ?

Reply:
yes, except you can either use inline ["data", "manifest"]  or just use a class constant.

Reply:
got it 👍